### PR TITLE
Maintain VST dev versions

### PR DESCRIPTION
**Pull request by QinshiWang**
This PR is supposed to provide an incremental test for https://github.com/coq/opam-coq-archive/pull/2260. Please DO NOT merge it.

### CODE REVIEW

**Comment by palmskog** [comment]
@QinshiWang @MSoegtropIMC here's what I think happens:
- the VST `.dev` packages are set to use exactly `coq-compcert.3.11`
- `coq-compcert.3.11` doesn't work with `menhir.dev`, even though the package doesn't rule out depending on `menhir.dev`
- compilation of VST `.dev` packages are never tested due to `coq-compcert.3.11` never building

@QinshiWang how about loosening up the CompCert requirement for the VST `.dev` packages? For example, you could allow `coq-compcert.dev`, and everything would likely work.

File: extra-dev/packages/coq-vst-32/coq-vst-32.dev/opam [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+]
+install: [
+  [make "install" "IGNORECOQVERSION=true" "ZLIST=platform" "BITSIZE=32"]
+]
+run-test: [
+  [make "-j%{jobs}%" "test" "IGNORECOQVERSION=true" "ZLIST=platform" "BITSIZE=32"]
+]
+depends: [
+  "coq" {>= "8.14" & < "8.17~"}
+  "coq-compcert" {= "3.11"}
```
Comment (palmskog) [inline extra-dev/packages/coq-vst-32/coq-vst-32.dev/opam:36]:
```suggestion
  "coq-compcert" {>= "3.11"}
```

File: extra-dev/packages/coq-vst/coq-vst.dev/opam [hunk 2]
```diff
@@ -1,44 +1,42 @@
...
+]
+run-test: [
+  [make "-j%{jobs}%" "test" "IGNORECOQVERSION=true" "ZLIST=platform" "BITSIZE=64"]
+]
 depends: [
-  "ocaml"
-  "coq" {>= "8.11.0"}
-  "coq-flocq" {>= "3.1.0" & < "4.0.0"} | "coq-flocq3" { = "dev"}
+  "coq" {>= "8.14" & < "8.17~"}
+  "coq-compcert" {= "3.11"}
```
Comment (palmskog) [inline extra-dev/packages/coq-vst/coq-vst.dev/opam:36]:
```suggestion
  "coq-compcert" {>= "3.11"}
```

**Review by palmskog — COMMENTED** [review]
Let's see what happens with compcert bound relaxed.

**Comment by palmskog** [comment]
@QinshiWang looks like there is some fundamental problem with the `coq-vst` `dev` packages:
```
# install -d "/builds/coq/opam-coq-archive/opam-root-4.07.1-2.1.2-sandbox/4.07.1/lib/coq/user-contrib/VST"
# for d in ./ atomics/ concurrency/ doc/ floyd/ msl/ progs64/ sepcomp/ veric/ zlist/; do install -d "/builds/coq/opam-coq-archive/opam-root-4.07.1-2.1.2-sandbox/4.07.1/lib/coq/user-contrib/VST/$d"; done
# for f in atomics/SC_atomics_base.v atomics/SC_atomics_base.vo atomics/verif_lock.v atomics/verif_lock.vo concurrency/cancelable_invariants.v concurrency/cancelable_invariants.vo concurrency/conclib.v concurrency/conclib.vo concurrency/ghosts.v concurrency/ghosts.vo concurrency/lock_specs.v concurrency/lock_specs.vo concurrency/semax_conc.v concurrency/semax_conc.vo concurrency/semax_conc_pred[...]
# install: cannot stat 'zlist/Zlength_solver.vo': No such file or directory
# install: cannot stat 'zlist/list_solver.vo': No such file or directory
# install: cannot stat 'zlist/sublist.vo': No such file or directory
# make: *** [Makefile:769: install] Error 1
```
Maybe we could separate all `dev` packages from actual releases to get the releases merged?

**Comment by QinshiWang** [comment]
I think `VST.dev` should allow both `compcert.3.11` and `compcert.dev`. The `dev` versions are always changing, there's no way to guarantee compatibility. It sounds to me like it should be set looser constraints.

**Comment by palmskog** [comment]
@QinshiWang indeed, I loosened up to `>= 3.11` in this not-to-be-merged PR, and the CI errors related to menhir disappeared. But then the errors related to `zlib` I mentioned in this thread appeared.

**Comment by QinshiWang** [comment]
Yes. I just saw it and I am looking into it.

**Comment by QinshiWang** [comment]
@palmskog I find a problem with `vst.2.10`. https://github.com/coq/opam-coq-archive/blob/master/released/packages/coq-vst/coq-vst.2.10/opam
I checked out `v2.10` of VST in Git and ran the `build` and `install` commands in the .opam file
```
make "-j%{jobs}%" "vst" "IGNORECOQVERSION=true" "BITSIZE=64"
make "install" "IGNORECOQVERSION=true" "BITSIZE=64"
```
I got a list of files not found
```
install: cannot stat 'progs64/append.vo': No such file or directory
install: cannot stat 'progs64/bin_search.vo': No such file or directory
install: cannot stat 'progs64/bst.vo': No such file or directory
install: cannot stat 'progs64/field_loadstore.vo': No such file or directory
...
```
But there's no errors or warnings when installing `2.10` with OPAM. Does OPAM do more than then `build` and `install` commands?

**Comment by QinshiWang** [comment]
The issue should have been fixed. But we need to wait until https://github.com/PrincetonUniversity/VST/pull/624 got merged. You might test by replacing
```
url {
  src: "git+https://github.com/PrincetonUniversity/VST.git#master"
}
```
with
```
url {
  src: "git+https://github.com/PrincetonUniversity/VST.git#opam"
}
```

File: extra-dev/packages/coq-vst-32/coq-vst-32.dev/opam [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  [make "-j%{jobs}%" "test" "IGNORECOQVERSION=true" "ZLIST=platform" "BITSIZE=32"]
+]
+depends: [
+  "coq" {>= "8.14" & < "8.17~"}
+  "coq-compcert" {>= "3.11"}
+  "coq-vst-zlist" {>= "2.11"}
+  "coq-flocq" {>= "4.1.0"}
+]
+url {
+  src: "git+https://github.com/PrincetonUniversity/VST.git#master"
```
Comment (palmskog) [inline extra-dev/packages/coq-vst-32/coq-vst-32.dev/opam:41]:
```suggestion
  src: "git+https://github.com/PrincetonUniversity/VST.git#opam"
```

File: extra-dev/packages/coq-vst/coq-vst.dev/opam [hunk 4]
```diff
@@ -1,44 +1,42 @@
...
+  "coq-compcert" {>= "3.11"}
+  "coq-vst-zlist" {>= "2.11"}
+  "coq-flocq" {>= "4.1.0"}
 ]
-synopsis: "Verified Software Toolchain"
-description:
-  "The software toolchain includes static analyzers to check assertions about your program; optimizing compilers to translate your program to machine language; operating systems and libraries to supply context for your program. The Verified Software Toolchain project assures with machine-checked proofs that the assertions claimed at the top of the toolchain really hold in the machine-language program, running in the operating-system context."
-flags: light-uninstall
 url {
   src: "git+https://github.com/PrincetonUniversity/VST.git#master"
```
Comment (palmskog) [inline extra-dev/packages/coq-vst/coq-vst.dev/opam:41]:
```suggestion
  src: "git+https://github.com/PrincetonUniversity/VST.git#opam"
```

**Review by palmskog — COMMENTED** [review]
@QinshiWang I'll see if I can replicate some error with 2.10 or not. But in the meantime, let's test with your `opam` branch in the CI here.

**Comment by palmskog** [comment]
Now `coq-vst.dev` is actually passing CI! However, `coq-vst-32.dev` still fails, but for a new reason:
```
# output-file          ~/opam-root-4.13.1-2.1.2-sandbox/log/coq-vst-32-4431-f294cf.out
### output ###
# cat: /builds/coq/opam-coq-archive/opam-root-4.13.1-2.1.2-sandbox/4.13.1/lib/coq/../coq-variant/compcert32/compcert/VERSION: No such file or directory
# cat: /builds/coq/opam-coq-archive/opam-root-4.13.1-2.1.2-sandbox/4.13.1/lib/coq/../coq-variant/compcert32/compcert/VERSION: No such file or directory
# Makefile:130: *** COMPCERT VERSION MISMATCH: COMPCERT_VERSION=version=3.11 but /builds/coq/opam-coq-archive/opam-root-4.13.1-2.1.2-sandbox/4.13.1/lib/coq/../coq-variant/compcert32/compcert/VERSION=.  Stop.
```

It looks like the package is using the bundled CompCert?

**Comment by QinshiWang** [comment]
I see. `coq-vst-32` needs to depend on `coq-compcert-32`. That was my bad. It should be fixed now.

**Comment by palmskog** [comment]
@QinshiWang unfortunately, the `coq-compcert-32.dev` package is not up to date - it requires the obsolete `coq-flocq3` package, and thus `coq-compcert-32.dev` fails.

However, since `coq-vst.dev` seems to work, I'm prepared to merge the original PR with the new changes applied, and we can take care of `coq-compcert-32.dev` later so that `coq-vst-32.dev` works.

**Comment by QinshiWang** [comment]
@palmskog I tested that `coq-vst-32.dev` works with `coq-compcert-32.3.11`.

**Pull request closed without merging**